### PR TITLE
fix(expo-image): reanimated on web

### PR DESCRIPTION
**Pull request by intergalacticspacehighway**
# Why

Reanimated on web mutates the [style](https://github.com/software-mansion/react-native-reanimated/blob/3188d92cc2fc0c44219f2cd9d2b09aea0b88e45c/src/reanimated2/js-reanimated/index.ts#L106) property of native DOM element using `ref`. This works as expected with React Native's `Image` component as it propogates `ref` to the underlying native element which is not the case the expo image. 

Repro snack - https://snack.expo.dev/@nishanbende/reanimated-web-and-expo-image-repro?platform=web

https://github.com/expo/expo/assets/23293248/537717eb-ea71-4eff-adba-f237dec39e5f

# How

Reanimated receives the `ref` of `Image` instance of class component and it returns as it doesn't find the [style](https://github.com/software-mansion/react-native-reanimated/blob/3188d92cc2fc0c44219f2cd9d2b09aea0b88e45c/src/reanimated2/js-reanimated/index.ts#L102) property on it. So we define a style property on the instance with custom setter and getter, this acts as proxy to the native container's style. Also, this means that reanimated can only modify the style property associated with the container view. But i guess that should be mostly fine.

# Test Plan
The above snack repro should be fixed. Reanimated triggered changes should be applied to the native view.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

# Alternate solution

We can expose [setNativeProps](https://github.com/software-mansion/react-native-reanimated/blob/3188d92cc2fc0c44219f2cd9d2b09aea0b88e45c/src/reanimated2/js-reanimated/index.ts#L95) on Image instance but as the comment suggests this is mostly deprecated and also a big downside is that reanimated will send us non-dom styles and we'll have to convert to DOM ourselves like [this](https://github.com/software-mansion/react-native-reanimated/blob/3188d92cc2fc0c44219f2cd9d2b09aea0b88e45c/src/reanimated2/js-reanimated/index.ts#L160) which would require undocumented imports from react-native-web!

### CODE REVIEW

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 1]
```diff
@@ -95,17 +95,17 @@ export default function ExpoImage({
...
   const imagePlaceholderContentFit = placeholderContentFit || 'scale-down';
   const imageHashStyle = {
     objectFit: placeholderContentFit || contentFit,
   };
-  const containerRef = React.useRef<HTMLDivElement | null>(null);
   const selectedSource = useSourceSelection(
     source,
     responsivePolicy,
-    containerRef,
+    containerViewRef as React.MutableRefObject<HTMLDivElement | null>,
```
Comment (intergalacticspacehighway) [inline packages/expo-image/src/ExpoImage.web.tsx:108]: felt bad doing this but universal types are not there yet 🤷‍♂️

**Review by tsapeta — APPROVED** [review]
Looks good to me, but let's see what @aleqsio thinks

File: yarn.lock [hunk 2]
```diff
@@ -1375,7 +1375,7 @@
   dependencies:
     eslint-visitor-keys "^3.3.0"
 
-"@eslint-community/regexpp@^4.10.0", "@eslint-community/regexpp@^4.5.1", "@eslint-community/regexpp@^4.6.1":
+"@eslint-community/regexpp@^4.10.0", "@eslint-community/regexpp@^4.6.1":
```
[the remark this answers is not in the document]
Comment (intergalacticspacehighway) [reply yarn.lock]: True, reverted. But i get these changes every time i run `yarn` in the project root. I am using yarn 1.22.4.

File: packages/expo-image/CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- Fix reanimated support on web by ([#29197](https://github.com/expo/expo/pull/29197) by [@nishan](https://github.com/intergalacticspacehighway))
```
Comment (lukmccall) [inline packages/expo-image/CHANGELOG.md:11]:
```suggestion
- Fix reanimated support on web. ([#29197](https://github.com/expo/expo/pull/29197) by [@nishan](https://github.com/intergalacticspacehighway))
```

**Review by lukmccall — COMMENTED** [review]
Looking good, but I think that we can implement that a little bit better. [Here is a PR ](https://github.com/software-mansion/react-native-reanimated/pull/4533/files) to the reanimated that adds ability to specify which ref should be use during animations. We can specify `getAnimatableRef` instead of defining `style` property on the `Image` component.

**Comment by intergalacticspacehighway** [comment]
> We can specify getAnimatableRef

Awesome. TIL about this API. I just tried and found that it doesn't work on web. This will require a fix in reanimated's web support 😅. I think [this line](https://github.com/software-mansion/react-native-reanimated/blob/3188d92cc2fc0c44219f2cd9d2b09aea0b88e45c/src/reanimated2/js-reanimated/index.ts#L84) needs to be replaced with `const component = viewRef._component.getAnimatableRef ? viewRef._component.getAnimatableRef() : viewRef._component;` to make it work. cc - @tomekzaw 🙏

**Comment by aleqsio** [comment]
I think the PR would be in an OK state to land it, but as @lukmccall said it might be better to use the reanimated API instead of adding this additional style logic.

**Comment by intergalacticspacehighway** [comment]
@aleqsio sounds good. Made a [PR](https://github.com/software-mansion/react-native-reanimated/pull/6088) on reanimated. Will update this PR when it gets merged.

**Comment by intergalacticspacehighway** [comment]
Thanks @lukmccall @aleqsio. Updated the PR to use `getAnimatableRef`, tested it locally. It will work in newer reanimated versions that includes https://github.com/software-mansion/react-native-reanimated/commit/fe0b50ea62dadbbf8c9888299a43ca3dca890e9c

**Review by aleqsio — APPROVED** [review]
Looks really good, thanks for moving to `getAnimatableRef`!

**Review by lukmccall — APPROVED** [review]

**Pull request merged**